biorbd_optim/optimal_control_program.py

```diff
@@ -459,11 +461,137 @@ def add_plot(self, fig_name, update_function, phase_number=-1, **parameters):
 
         nlp["plot"][plot_name] = custom_plot
 
+    def export_eocar_ode_model(self):
+        from acados_template import AcadosModel
+        from casadi import MX, Function, external, SX, vertcat, sin, cos
+
+
+        m = biorbd.Model(self.nlp[0]['model'].path().relativePath().to_string())
+        model_name = "model_name"
+
+        # Declare model variables
+        x = MX.sym('x', m.nbQ() * 2)
+        u = MX.sym('u', m.nbQ())
+        xdot = MX.sym('dx', m.nbQ() * 2)
+
+        f = Function('f', [x, u],
+                     [vertcat(x[m.nbQ():], m.ForwardDynamics(x[:m.nbQ()], x[m.nbQ():], u).to_mx())]).expand()
+        x = SX.sym('x', m.nbQ() * 2)
+        u = SX.sym('u', m.nbQ())
+        xdot = SX.sym('dx', m.nbQ() * 2)
+        f_expl = f(x, u)
+        f_impl = xdot - f_expl
+
+        model = AcadosModel()
+        model.f_impl_expr = f_impl
+        model.f_expl_expr = f_expl
+        model.x = x
+        model.xdot = xdot
+        model.u = u
+        # model.z = z
+        model.p = []
+        model.name = model_name
+
+        return model
+
+    def __prepare_acados(self):
+        from acados_template import AcadosOcp, AcadosOcpSolver, AcadosModel
+        import numpy as np
+        import scipy.linalg
+        from casadi import Function, external, SX, sin, cos
+
+        os.environ["ACADOS_SOURCE_DIR"] = "/home/dangzilla/Documents/Programmation/acados"
+        # create ocp object to formulate the OCP
+        ocp = AcadosOcp()
+
+        # # set model
+        model = self.export_eocar_ode_model()
+        ocp.model = model
+
+        # set time
+        ocp.solver_options.tf = self.nlp[0]["tf"]
+
+        # set dimensions
+        ocp.dims.nx = self.nlp[0]["nx"]
+        ocp.dims.nu = self.nlp[0]["nu"]
+        ocp.dims.ny = ocp.dims.nx + ocp.dims.nu
+        ocp.dims.ny_e = self.nlp[0]["nx"]
+        ocp.dims.N = self.nlp[0]["ns"]
+
+        # set cost module
+        ocp.cost.cost_type = 'LINEAR_LS'
+        ocp.cost.cost_type_e = 'LINEAR_LS'
+
+        # set weight à modifier avec objective functions (user needs to define this TODO)
+        Q = 0.00 * np.eye(ocp.dims.nx)
+        R = 5 * np.eye(ocp.dims.nu)
+
+        ocp.cost.W = scipy.linalg.block_diag(Q, R)
+
+        ocp.cost.W_e = Q
+
+        # set Lagrange term matrices
+        ocp.cost.Vx = np.zeros((ocp.dims.ny, ocp.dims.nx))
+        ocp.cost.Vx[:ocp.dims.nx, :] = np.eye(ocp.dims.nx)
+
+        # Vu = np.zeros((ny, nu))
+        # Vu[nx:, :] = np.eye(nu)
+        # ocp.cost.Vu = Vu
+
+        Vu = np.zeros((ocp.dims.ny, ocp.dims.nu))
+        Vu[ocp.dims.nx:, :] = np.eye(ocp.dims.nu)
+        ocp.cost.Vu = Vu
+
+        # set Mayer term matrices
+        ocp.cost.Vx_e = np.zeros((ocp.dims.nx, ocp.dims.nx))
+
+        ocp.cost.yref = np.zeros((ocp.dims.ny,))
+        ocp.cost.yref_e = np.ones((ocp.dims.ny_e,))
+
+        # set constraints
+        ocp.constraints.x0 = np.array(self.nlp[0]["X_bounds"].min[:, 0])
```

Comment:
I remember the nomenclature of ACADOS to be confusing with this. If x0 is indeed the constraint for initial value (and not the initial guess as the name suggests), then you did the right thing. However, please raise an error if `min` and `max` are not equal (since it seems that you are assuming that the initial values are set and fixed while there is nothing in theory that forces the user to not allow free variables here)